### ISSUE TITLE
Few issues connecting a generic_android device

### ISSUE THREAD

**Issue by Jerome-Neanne** [ARM-software/workload-automation#1035]
I had 2 issues when connecting NXP imx8 generic android:

Issue#1:
========================================
INFO     Removing existing output directory.
INFO     Creating output directory.
INFO     Initializing run
INFO     Initializing execution context
INFO     Connecting to target
adb server is out of date.  killing...
* daemon started successfully *
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/wa/framework/signal.py", line 323, in wrap
ERROR        yield
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/wa/framework/target/manager.py", line 143, in _init_target
ERROR        self.target.connect()
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/target.py", line 1195, in connect
ERROR        super(AndroidTarget, self).connect(timeout=timeout, check_boot_completed=check_boot_completed)
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/target.py", line 273, in connect
ERROR        self._update_modules('connected')
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/target.py", line 837, in _update_modules
ERROR        self._install_module(mod, **params)
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/target.py", line 849, in _install_module
ERROR        mod.install(self, **params)
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/module/__init__.py", line 60, in install
ERROR        setattr(target, attr_name, cls(target, **params))
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/module/hwmon.py", line 137, in __init__
ERROR        self.scan()
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/module/hwmon.py", line 140, in scan
ERROR        values_tree = self.target.read_tree_values(self.root, depth=3, tar=True)
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/target.py", line 777, in read_tree_values
ERROR        strip_null_chars)
ERROR      File "/home/nxf44606/.local/lib/python3.5/site-packages/devlib/target.py", line 714, in read_tree_tar_flat
ERROR        tar_buf = gzip.GzipFile(fileobj=tar_gz_bytes).read()
ERROR      File "/usr/lib/python3.5/gzip.py", line 274, in read
ERROR        return self._buffer.read(size)
ERROR      File "/usr/lib/python3.5/gzip.py", line 461, in read
ERROR        if not self._read_gzip_header():
ERROR      File "/usr/lib/python3.5/gzip.py", line 409, in _read_gzip_header
ERROR        raise OSError('Not a gzipped file (%r)' % magic)
ERROR    
ERROR    OSError(Not a gzipped file (b'~)'))
INFO     Done.
INFO     Run duration: N/A
INFO     Ran a total of 0 iterations: 
INFO     Results can be found in wa_output
WARNING  There were errors during execution.
WARNING  Please see wa_output/run.log
=======================================

I tried WA indicated here: https://github.com/ARM-software/workload-automation/issues/1027
augmentations: [~~]   # disable default augmentations

But this has no effect. 

Then I set adb root on the server side to resolve. Then I hit issue#2.

Issue#2:
===========================================
INFO     Removing existing output directory.
INFO     Creating output directory.
INFO     Initializing run
INFO     Initializing execution context
INFO     Connecting to target
adb server is out of date.  killing...
* daemon started successfully *
ERROR    Got exit code 1
ERROR    from target command: PATH=/data/local/tmp/bin:$PATH && if [ -e /sys/devices/system/cpu/cpu1/online ]; then echo 1; else echo 0; fi
ERROR    OUTPUT: su: invalid uid/gid '-c'
ERROR    
INFO     Done.
INFO     Run duration: N/A
INFO     Ran a total of 0 iterations: 
INFO     Results can be found in wa_output
WARNING  There were errors during execution.
WARNING  Please see wa_output/run.log
==========================================

Basically this is because "adb shell su -c" is not supported on my platform. "-c" is causing the issue. Just removing the -c just works fine so I'm wondering why this is needed. Can't find much description about this option.

To solve this I made a quick hack in devlib/utils/android.py 
    #parts += ['shell', JN
    #          command if not as_root else 'su -c {}'.format(quote(command))]
    parts += ['shell',
              command]

After that I can run my agenda successfully. Just wondering why I have to perform this hack. Is something not "android_generic" on my platform?
Is their any option I missed to avoid that?

**Comment by marcbonnici** [comment]
Hello,

Issue 1:
This looks like an issue with the `hwmon` devlib "module" rather than an augmentation, in the latest version of devlib there should now be a better error message to indicate which module is failing to install as understandably the error you received is not very easy to parse. 
The method to disable the module if you do not wish to use adb root is detailed here: https://workload-automation.readthedocs.io/en/latest/faq.html#q-i-get-the-error-module-x-failed-to-install-on-target. 

Issue 2:
The `su -c` is designed to execute the parameter with root privileges however not all versions of `su` support this so devlib should automatically fall back to piping into su as of https://github.com/ARM-software/devlib/commit/dc5f4c6b499a22429681b4bd10e3902287b042f7.
If this is not functioning correctly would you be able to post the version of devlib you are using and your full `run.log`?

**Comment by Jerome-Neanne** [comment]
Hello,

Thanks for such responsiveness!
find attached the full log.
[run.log](https://github.com/ARM-software/workload-automation/files/3845447/run.log)

I had a look to devlib official repo and clearly see that my code is not aligned on the latest version.
From what I see in utils/version.py, my version is 1.1.2
Just realized that I did not made venv activate since last reboot I might have some python version conflict. I'll clarify and come back to you.

**Comment by Jerome-Neanne** [comment]
For issue#1, I tested the method you pointed do disable modules (load_default_modules: False) and confirm this is working perfectly.

For issue#2: No difference running within venv when I'm running with adb root still failling due tu su -c

and I'm using same devlib version 1.1.2
[run.log](https://github.com/ARM-software/workload-automation/files/3845755/run.log)

**Comment by marcbonnici** [comment]
Ok great. 

The fixes for failing with  `su -c` have added after our last official release of v1.1.2, would you be able to install the latest version of devlib from github and see if that resolves the problem?

**Comment by Jerome-Neanne** [comment]
Sure. I'll try this later today

**Comment by Jerome-Neanne** [comment]
I confirm latest devlib is fixing issue#2:
in venv:
pip uninstall devlib
pip install git+https://github.com/ARM-software/devlib.git

Thanks!
